Add IP comparison on machineset test

rlobillo opened the pull request:
Covers automation for https://bugzilla.redhat.com/show_bug.cgi?id=2022627

Due to the work done to fix the previous BZ, the Machine Status section now includes all the addresses defined in the OSP instances.

This change adds this check for workers created by a machineset in the existing test "[sig-installer][Feature:openstack] The OpenStack platform on instance creation should follow machineset specs [Suite:openshift/conformance/parallel]".

rlobillo commented:
/cc @mdbooth

pierreprinetti commented on test/extended/openstack/servers.go:
```diff
@@ -76,10 +79,14 @@ var _ = g.Describe("[sig-installer][Feature:openstack] The OpenStack platform",
 					o.Expect(instanceFlavor.Name).To(o.Equal(flavorMachineSet), "Flavor not matching for instance %q", instance.Name)
 					o.Expect(instanceImage.Name).To(o.Equal(imageMachineSet), "Image not matching for instance %q", instance.Name)
 					o.Expect(instanceSgs).To(o.Equal(sgMachineSets), "SGs not matching for %q", instance.Name)
+					o.Expect(instanceAddresses).To(o.Equal(addressesMachine), "Addresses not matching for instance %q", instance.Name)
```
Can you please mention the BZ in a comment right above this new check?

pierreprinetti replied:
Also, depending on your personal taste, the four lines you added can be replaced with this one-liner:

```Go
o.Expect(parseInstanceAddresses(instance.Addresses)).To(o.Equal(getAddressesFromMachine(machine))
```

the advantage would be that we have all dependent code packed together. Fewer preparatory steps above means that while reading, if I am only interested in `sgMachineSets` for example, I don't have to read about `instanceAddresses`.

but it's really up to you.

pierreprinetti commented:
I am mildly concerned by this proliferation of checks in a single test. Do you think we should continue checking all MachineSet and Machine properties in a single test, or shall we split each check in its own test?

pierreprinetti commented on test/extended/openstack/servers.go:
```diff
@@ -172,13 +179,43 @@ func parseInstanceSgs(securityGroups []map[string]interface{}) []string {
+
+	for _, entry := range strings.Split(s, " ") {
+		if strings.HasPrefix(entry, "addr") {
+			addr := net.ParseIP(strings.Split(entry, ":")[1])
```
In case there is no `:`, the slice resulting from `strings.Split(entry, ":")` will have length `1` and indexing `[1]` will panic.

Also, in order to account for IPv6, we should probably use [`strings.SplitN`](https://pkg.go.dev/strings#SplitN):

```Go
	if strings.HasPrefix(entry, "addr") {
		addrLine := strings.SplitN(entry, ":", 2)
		if len(addrLine) > 1 {
			addr := net.ParseIP(addrLine[1])
			if addr == nil {
				return nil, fmt.Errorf("invalid ip format: %v", addrLine[1])
			}
			result = append(result, addr.String())
		}
	}
```

Alternatively, we can leverage [`strings.TrimPrefix`](https://pkg.go.dev/strings#TrimPrefix). `TrimPrefix` returns the string unchanged if the prefix was not found:
```Go
	if addrValue := strings.TrimPrefix(entry, "addr:"); addrValue != entry {
		addr := net.ParseIP(addrValue)
		if addr == nil {
			return nil, fmt.Errorf("invalid ip format: %v", addrValue)
		}
		result = append(result, addr.String())
	}
	return result, nil
```

rlobillo replied:
Thanks!! Done

rlobillo commented:
> I am mildly concerned by this proliferation of checks in a single test. Do you think we should continue checking all MachineSet and Machine properties in a single test, or shall we split each check in its own test?

Agree, I moved it to a different testcase, so we are checking the addresses for all the machines, master, workers, belonging to a machine set or not.

rlobillo commented:
/retest

pierreprinetti commented:
/retest-required

pierreprinetti commented:
/approve
/lgtm

Great work, thanks!
